chainer/link_hooks/spectral_normalization.py

```diff
@@ -0,0 +1,262 @@
+
+    def added(self, link):
+        # Define axis and register ``u`` if the weight is initialized.
+        if not hasattr(link, self.weight_name):
+            raise ValueError(
+                "Weight '{}' does not exist!".format(self.weight_name))
+        if isinstance(link, (L.Deconvolution2D, L.DeconvolutionND)):
+            self.axis = 1
+        if getattr(link, self.weight_name).array is not None:
+            self._prepare_parameters(link)
+
+    def deleted(self, link):
+        # Remove approximate vector ``u`` and parameter ``gamma` if exists.
+        delattr(link, self.vector_name)
+        if self.use_gamma:
+            del link.gamma
+
+    def forward_preprocess(self, cb_args):
+        # This method normalizes target link's weight spectrally
+        # using power iteration method
+        link = cb_args.link
+        input_variable = cb_args.args[0]
+        if not self._initialied:
+            self._prepare_parameters(link, input_variable)
+        weight = getattr(link, self.weight_name)
+        # For link.W or equivalents to be chainer.Parameter
+        # consistently to users, this hook maintains a reference to
+        # the unnormalized weight.
+        self.original_weight = weight
+        # note: `normalized_weight` is ~chainer.Variable
+        normalized_weight = self.normalize_weight(link, weight)
+        setattr(link, self.weight_name, normalized_weight)
+
+    def forward_postprocess(self, cb_args):
+        # Here, the computational graph is already created,
+        # we can reset link.W or equivalents to be Parameter.
+        link = cb_args.link
+        setattr(link, self.weight_name, self.original_weight)
+
+    def _prepare_parameters(self, link, input_variable=None):
+        """Prepare one buffer and one parameter.
+
+        Args:
+            link (:class:`~chainer.Link`): Link to normalize spectrally.
+            input_variable (:class:`~chainer.Variable`):
+                The first minibatch to initialize weight.
+
+        """
+        if getattr(link, self.weight_name).array is None:
+            if input_variable is not None:
+                link._initialize_params(input_variable.shape[1])
+        initialW = getattr(link, self.weight_name)
+        if initialW.shape[self.axis] == 0:
+            raise ValueError(
+                'Expect {}.shape[{}] > 0'.format(self.weight_name, self.axis)
+            )
+        u = link.xp.random.normal(
+            size=(initialW.shape[self.axis],)).astype(dtype=initialW.dtype)
+        setattr(link, self.vector_name, u)
+        link.register_persistent(self.vector_name)
+        if self.use_gamma:
+            # Initialize the scaling parameter with the max singular value.
+            weight_matrix = self.reshape_W(initialW.array)
+            _, s, _ = link.xp.linalg.svd(weight_matrix)
+            gamma_shape = [1] * initialW.ndim
+            with link.init_scope():
+                link.gamma = variable.Parameter(s[0], gamma_shape)
```

Comment:
`gamma_shape` can be `()`.

Reply:
Really?
What layer has the weight of 0-dim?

Reply:
What I wrote was ambiguous.  I meant: use (automatic) broadcast later and
```suggestion
                link.gamma = variable.Parameter(s[0], ())
```